core/src/test/scala/unit/kafka/log/CleanerTest.scala

```diff
@@ -610,11 +611,75 @@ class CleanerTest extends JUnitSuite {
     assertEquals(-1, map.get(key(4)))
   }
 
+  @Test
+  def testCleanCorruptMessageSet() {
+    // This test verifies that messages corrupted by KAFKA-4298 are fixed by the cleaner
+
+    val log = makeLog()
+    val cleaner = makeCleaner(10)
+
+    val kvs1 = (0 until 2) zip (0 until 2)
+    val kvs2 = (3 until 5) zip (3 until 5)
+
+    writeInvalidCleanedMessage(log, 25, kvs1 ++ kvs1) // one compressed entry with dups
+    writeInvalidCleanedMessage(log, 50, kvs2) // and one without (should still be fixed by the cleaner)
+
+    log.roll()
+
+    cleaner.clean(LogToClean(TopicAndPartition("test", 0), log, 0, log.activeSegment.baseOffset))
+
+    for (segment <- log.logSegments; shallowMessage <- segment.log.iterator; deepMessage <- ByteBufferMessageSet.deepIterator(shallowMessage))
+      assertEquals(shallowMessage.message.magic, deepMessage.message.magic)
```

Comment:
Do we need to check the offsets as well? It wasn't clear to me if we need to do something extra for that case.